pyslim/slim_tree_sequence.py

```diff
@@ -490,6 +490,139 @@ def recapitate(self,
+        ``recombination_rate`` or a ``recombination_map`` are provided, there will be 
+        *no* recombination.
+
+        However, if ``recombination_rate`` *is* provided, then recapitation will
+        use a constant rate of recombination on a discretized map -- in other words,
+        recombinations in the coalescent portion of the simulation will only occur
+        at integer locations, just as in SLiM. If you do not want this to happen,
+        you need to construct a ``recombination_map`` explicitly.
+
+        Note that ``Ne`` defaults to ``1.0``; you probably
+        want to set it explicitly.
+
+        :param int slim_time: The desired period of time to simulate in msprime,
+            in units of generations
+        :param int Ne: The effective population size for :meth:`msprime.simulate`,
+            in units of diploid individuals
+        :param int samples: Sample count for :meth:`msprime.simulate`,
+            in units of haploid samples,
+        :param float mutation_rate: A (constant) mutation rate,
+            in units of mutations per nucleotide per unit of time.
+        :param float recombination_rate: A (constant) recombination rate,
+            in units of crossovers per nucleotide per unit of time.
+        :type recombination_map: :class`msprime.RecombinationMap`
+        :param recombination_map: The recombination map, or None,
+            if recombination_rate is specified.
+        :param dict kwargs: Any other arguments to :meth:`msprime.simulate`.
+        '''
+        if "keep_first_generation" in kwargs:
+            raise ValueError("The keep_first_generation argument is deprecated:"
+                             "the FIRST_GEN flag is no longer used.")
+
+        discrete_msprime = hasattr(msprime, "RateMap")
+
+        if recombination_rate is not None:
+            if recombination_map is not None:
+                raise ValueError("Cannot specify length/recombination_rate along with a recombination map")
+            if discrete_msprime:
+                recombination_map = msprime.RecombinationMap(positions = [0.0, self.sequence_length],
+                                                             rates = [recombination_rate, 0.0])
+            else:
+                recombination_map = msprime.RecombinationMap(positions = [0.0, self.sequence_length],
+                                                             rates = [recombination_rate, 0.0],
+                                                             num_loci = int(self.sequence_length))
+        if discrete_msprime and ('discrete_genome' not in kwargs):
+            kwargs['discrete_genome'] = True
+
+        slim_indivs = self.individuals_alive_at(0)
+        slim_nodes = []
+        for ind in slim_indivs:
+            slim_nodes.extend(self.individual(ind).nodes)
+        slim_nodes = np.array(slim_nodes)
+    
+        if samples is None:
+            samples=len(slim_nodes)
+        assert(len(slim_nodes) == samples)
```

Comment:
I think you meant
```
if samples is None:
    samples = np.arange(len(slim_nodes))
```

Also, I don't think there's any need for `assert len(slim_nodes) == samples`? We could continue the simulation and only be interested in a small number of lineages, right? Or am I missing something?

Reply:
To the best of my understanding, the number of individuals alive at time=0 in the slim tree sequence must be equal to the sample size simulated in msprime, as we map the two sets of nodes together. I may be missing something though! Or alternatively we could map msprime lineages to arbitrary slim individuals. Not sure how much control to exert over this within the method, but for those two sizes to be different would require a more general use way of linking nodes I believe. 

This is derived from the following lines from the tutorial - i think the variable samples should be renamed sample_size though, a bit confusing there.
```
slim_indivs = rts.individuals_alive_at(0)
slim_nodes = []
for ind in slim_indivs:
   slim_nodes.extend(ts.individual(ind).nodes
slim_nodes = np.array(slim_nodes)
assert(len(slim_nodes) == 20000)
```

Reply:
Or to clarify - we already do map msprime lineages to random slim nodes, but I guess removing the assertion allows for there to be more slim nodes than msprime nodes at the merger. We would have to set `replace=True` for `np.random.choice`if the msprime sample size were set larger than slim nodes alive at 0, which would map multiple msprime nodes to some slim nodes and would not guarantee full coverage of slim lineages in the merger.

Reply:
Hm, right - this probably needs more explaining in the vignette. To make things concrete, let's say we run a SLiM simulation with a population size of 100 for a while, and then follow it up with 50 more generations of msprime simulation with 100 samples from an Ne of 100. So, the end of the SLiM simulation will be "50 generaitons ago", and what we're doing is we (a) start with 100 lineages, today; (b) run these back through time until 50 generations ago, and (c) match them up, randomly, with the individuals alive at the end of the SLiM simulation. In step (c), we match them *randomly* because under a neutral model, any of those individuals are as likely to be ancestors as any other ones. Note that the result of step (b) will be *fewer* than 100 lineages - probably quite a lot fewer at any particular locus - and so when we match them up, some SLiM individuals don't get matched to lineages. These are individuals that didn't leave behind any genetic material that made it until today (at the locus we're looking at). So, there's no problem having fewer lineages than SLiM individuals. Also note that there's no constraint in msprime regarding the number of samples and Ne - you could take a million samples from a population with Ne=100 if you wanted - this might represent juveniles from a population with very high juvenile mortality, for instance.

So, the only awkwardness I see is if we end up with *more* lineages at the end of the SLiM simulation than individuals. But I'm not sure we should disallow this, as maybe someone would have a good reason for doing things that way?

Reply:
That clarification helps a lot - seems like a non-issue overall, just requires some documentation. In the rare case of msprime having a higher count of nodes at the merger time, do you think it is better to a) prevent this case using an assertion or b) allow it by conditionally removing the replacement condition in `np.random.choice()`?

Reply:
Hm - thinking this through -`replace=False` is like saying that the msprime lineages *are the same individuals as* the final SLiM chromosomes; while `replace=True` is like saying there's a round of WF reproduction between the two. If the SLiM population size is much bigger than the number of lineages, the difference will be tiny. I like the `replace=False` interpretation better, but as you say, `replace=True` lets the number of lineages be greater than the population size (as you'd get from a sudded population size change in a WF model). 

So, I agree, changing this to `replace=True` is a good way to do it.

Reply:
I believe I found a better solution:

```
new_nodes = np.where(new_ts.tables.nodes.time == time)[0]

    if len(new_nodes) > len(old_nodes):
        new_nodes = np.random.choice(new_nodes, len(old_nodes), replace=False)
    node_map = np.repeat(tskit.NULL, new_ts.num_nodes)
    node_map[new_nodes] = np.random.choice(old_nodes, len(new_nodes), replace=False)
```

Basically, if the number of lineages is greater than the population size I drop extra lineages randomly and have an equal amount of lineages to the population size. How does that look to you? Either solution could work I think, this one just preserves a one-to-one ratio of lineages between the tree sequences.

Reply:
Hm, this is a bit tricky. It doesn't work to drop lineages, since that would imply that some of the samples in the resulting tree sequence don't inherit from *anyone* past this time.  Also, `node_mapping` is - I think  - required to be one-to-one, so that we can't have two lineages mapipng to a single individual (but [this needs figuring out](https://github.com/tskit-dev/tskit/blob/c9b2c920b6c4e1aaebcbc91dc4c6e13e919bdd08/c/tskit/tables.c#L9400) - and if so, `replace=True` isn't allowed.

Reply:
Ah that makes sense. I think the solution for now may have to be to disallow it then. Having `replace=True` sets a not great precedent and could cause issues down the road, and yes having samples that don't inherit from anyone past that time is wrong.

Reply:
It seems unlikely to be a restriction: if the number of samples is less than or equal to the size of the final SLiM population, then we'd never have too many samples, anyhow.